### PR TITLE
router: allow propagating attempt count in header

### LINKED ISSUE #4210: Propagate attempt count in request headers
We're interested in having the upstream attempt count be included in the request headers, e.g. something like `-x-envoy-attempt-count` which starts at 1 with the first request and is incremented once per retry. Our motivation for this is to provide this information to some of our existing systems that rely on this for their caching strategy. 

Seems like this was also [proposed/implemented for gRPC](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exposed-retry-metadata), so it's probably a generally useful feature.

In terms of implementation I imagine this being a simple incrementing value maintained in the `Router`. Not completely sure where this would be configured, `VirtualHost` maybe?

### PR DESCRIPTION
**Pull request by snowp**
Adds the ability to configure x-envoy-attempt-count to be set to the attempt count to allow upstreams to infer how many times the request has been retried.

*Risk Level*: Medium
*Testing*: Integration test
*Docs Changes*: Updated the header documentation to include new header
*Release Notes*: Added release note
Fixes #4210

### CODE REVIEW

File: test/integration/integration_test.cc [hunk 1]
```diff
@@ -109,6 +109,8 @@ TEST_P(IntegrationTest, RouterUpstreamResponseBeforeRequestComplete) {
 
 TEST_P(IntegrationTest, Retry) { testRetry(); }
 
+TEST_P(IntegrationTest, RetryAttemptCount) { testRetryAttemptCountHeader(); }
```
Comment (junr03) [inline test/integration/integration_test.cc]: I think we are lacking unit test on making sure the configuration is wired up correctly, probably in test/common/router/config_impl_test.cc.

Less clear to me is if we should also have unit testing at the router level in test/common/router/router_test.cc

File: docs/root/intro/version_history.rst [hunk 2]
```diff
@@ -77,6 +77,7 @@ Version history
 * rbac network filter: a :ref:`role-based access control network filter <config_network_filters_rbac>` has been added.
 * rest-api: added ability to set the :ref:`request timeout <envoy_api_field_core.ApiConfigSource.request_timeout>` for REST API requests.
 * router: added ability to set request/response headers at the :ref:`envoy_api_msg_route.Route` level.
+* router: addded ability to set attempt count in upstream requests, see :ref:`envoy_api_field_route.VirtualHost.include_request_attempt_count`.
```
Comment (dio) [inline docs/root/intro/version_history.rst]: nit: s/addded/added.

And also this introduces warning, 

```
Warning, treated as error:
/../generated/rst/intro/version_history.rst:80:undefined label: envoy_api_field_route.virtualhost.include_request_attempt_count (if the link has no caption the label must precede a section header)
```

Probably something like the following might work:

```
* router: added ability to set attempt count in upstream requests, see :ref:`virtual host's include request attempt count flag <envoy_api_field_route.VirtualHost.include_request_attempt_count>`.
```

**Comment by junr03** [comment]
looks good to me @ggreenway can you take a look?

**Comment by ggreenway** [comment]
This is not my area of expertise. @alyssawilk can you take this one?

**Comment by ggreenway** [comment]
@snowp you have merge conflict :(

**Comment by mattklein123** [comment]
@alyssawilk is out today. I can take it.

File: api/envoy/api/v2/route/route.proto [hunk 3]
```diff
@@ -109,6 +109,10 @@ message VirtualHost {
   // specific; see the :ref:`HTTP filter documentation <config_http_filters>`
   // for if and how it is utilized.
   map<string, google.protobuf.Struct> per_filter_config = 12;
+
+  // Decides whether the x-envoy-attempt-count header should be included in upstream
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: Can you add to https://www.envoyproxy.io/docs/envoy/latest/configuration/http_filters/router_filter#http-headers-set and cross link?

File: include/envoy/router/router.h [hunk 4]
```diff
@@ -612,7 +617,12 @@ class RouteEntry : public ResponseEntry {
    */
   template <class Derived> const Derived* perFilterConfigTyped(const std::string& name) const {
     return dynamic_cast<const Derived*>(perFilterConfig(name));
-  }
+  };
+
+  /**
+   * @return bool whether x-envoy-attempt-count should be included on the upstream request.
+   */
+  virtual bool includeAttemptCount() const PURE;
```
Comment (mattklein123) [inline include/envoy/router/router.h]: I only see configuration for virtual host. Is this used? Do we need route configuration also? If we don't want this now can you comment that this is pulled from the virtual host config currently?

File: source/common/router/router.cc [hunk 5]
```diff
@@ -287,6 +287,11 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
     headers.removeEnvoyUpstreamRequestTimeoutAltResponse();
   }
 
+  include_attempt_count_ = route_entry_->includeAttemptCount();
+  if (include_attempt_count_) {
+    headers.insertEnvoyAttemptCount().value(1);
```
Comment (mattklein123) [inline source/common/router/router.cc]: nit: use `attempt_count_` here which you initialize to 1?

File: docs/root/configuration/http_filters/router_filter.rst [hunk 6]
```diff
@@ -22,6 +22,16 @@ ingress/response path. They are documented in this section.
 .. contents::
   :local:
 
+.. _config_http_filters_router_x-envoy-attempt-count:
```
Comment (mattklein123) [inline docs/root/configuration/http_filters/router_filter.rst]: I think this should be in the "set" section below vs. the "consumed" section?

File: api/envoy/api/v2/route/route.proto [hunk 7]
```diff
@@ -110,7 +110,9 @@ message VirtualHost {
   // for if and how it is utilized.
   map<string, google.protobuf.Struct> per_filter_config = 12;
 
-  // Decides whether the x-envoy-attempt-count header should be included in upstream
+  // <config_http_filters_router_x-envoy-original-path>` header.
+  // Decides whether the :ref:`x-envoy-attempt-count
+  // <config_http_filters_router_x-envoy-attempt-count>` header should be included in upstream
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: nit: s/in/in the

**Comment by mattklein123** [comment]
@snowp sorry needs a master merge.

File: api/envoy/api/v2/route/route.proto [hunk 8]
```diff
@@ -109,6 +109,12 @@ message VirtualHost {
   // specific; see the :ref:`HTTP filter documentation <config_http_filters>`
   // for if and how it is utilized.
   map<string, google.protobuf.Struct> per_filter_config = 12;
+
+  // <config_http_filters_router_x-envoy-original-path>` header.
+  // Decides whether the :ref:`x-envoy-attempt-count
+  // <config_http_filters_router_x-envoy-attempt-count>` header should be included
+  // in the upstream request. Defaults to false.
+  bool include_request_attempt_count = 14;
```
Comment (lizan) [inline api/envoy/api/v2/route/route.proto]: nit: update next free id in comment above.

File: api/envoy/api/v2/route/route.proto [hunk 9]
```diff
@@ -109,6 +109,12 @@ message VirtualHost {
   // specific; see the :ref:`HTTP filter documentation <config_http_filters>`
   // for if and how it is utilized.
   map<string, google.protobuf.Struct> per_filter_config = 12;
+
+  // <config_http_filters_router_x-envoy-original-path>` header.
```
Comment (lizan) [inline api/envoy/api/v2/route/route.proto]: is this related to `x-envoy-original-path`?

File: api/envoy/api/v2/route/route.proto [hunk 10]
```diff
@@ -109,6 +109,12 @@ message VirtualHost {
...
+  // Decides whether the :ref:`x-envoy-attempt-count
```
Comment (lizan) [inline api/envoy/api/v2/route/route.proto]: How this play with `suppress_envoy_headers` in router.proto?

File: source/common/router/router.cc [hunk 11]
```diff
@@ -287,6 +287,11 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
     headers.removeEnvoyUpstreamRequestTimeoutAltResponse();
   }
 
+  include_attempt_count_ = route_entry_->includeAttemptCount();
+  if (include_attempt_count_) {
+    headers.insertEnvoyAttemptCount().value(attempt_count_);
```
[the remark this answers is not in the document]
Comment (snowp) [reply source/common/router/router.cc]: That sounds reasonable, I'll update the documentation

File: docs/root/intro/version_history.rst [hunk 12]
```diff
@@ -3,6 +3,8 @@ Version history
 
 1.9.0 (pending)
 ===============
+* router: added ability to set attempt count in upstream requests, see :ref:`virtual host's include request
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: alpha order.

File: api/envoy/api/v2/route/route.proto (hunk 10 above)
Comment (lizan) [reply api/envoy/api/v2/route/route.proto]: I'm fine either way, perhaps document the behavior when both are true? @htuch any suggestion?

**Comment by snowp** [comment]
@mattklein123 ptal
